src/engine/renderer/tr_shader.cpp

```diff
@@ -41,6 +41,7 @@ static char          *s_shaderText;
 // dynamically allocated memory if it is valid.
 static shaderTable_t table;
 static std::array<shaderStage_t, MAX_SHADER_STAGES> stages;
+static uint8_t numStages;
```

Comment:
This data is only used when processing the global temporary `stages` array.

I don't see why that global temporary `numStage` variable would have more or less less risk to be out of sync with the global temporary `stages` array, if stored or not in the global temporary `shader` struct. It was global, it is still global. The risks are the same. The code behaviour of having thousands of lines modifying global variables before calling a function to copy their content to a newly-allocated “permanent” shader is left as is.

Why not making a global struct for all the temporary things? Like this:

```c
struct tmpShaderData_t {
	shaderTable_t table;
	std::array<shaderStage_t, MAX_SHADER_STAGES> stages;
	shader_t shader;
	texModInfo_t texMods[ MAX_SHADER_STAGES ][ TR_MAX_TEXMODS ];
};

static tmpShaderData_t tmp;
```

It also means we could do:

```c
static void ClearGlobalShader()
{
	ResetStruct( tmp );
}
```

It would also makes more obvious this kind of code:

```c
for ( uint8_t s = 0; s < numStages; s++ )
{
	shaderStage_t *stage = &stages[ s ];

	if ( shader.numDeforms > 0 ) {
		stage->deformIndex = gl_shaderManager.getDeformShaderIndex( shader.deforms, shader.numDeforms );
	}
}
```

by turning it into:

```c
for ( uint8_t s = 0; s < tmp.numStages; s++ )
{
	shaderStage_t *stage = &tmp.stages[ s ];

	if ( tmp.shader.numDeforms > 0 ) {
		tmp.stage->deformIndex = gl_shaderManager.getDeformShaderIndex( tmp.shader.deforms, tmp.shader.numDeforms );
	}
}
```

I'm personnally annoyed by all the global temporary variables being named `shader` and `stages` so I would even love that. I see no reason to use the shader and stage structs for storing data that is only used at parsing them and never used at rendering time, we have some others of them, for example shader variables that are only used to configure stages at load time.

This change would be probably too intrusive so I may postpone it to another PR, but that is something I consider.

Reply:
> This accounting variable already existed

How do you mean?

I think `table` is not temporary, but other than that the `tmp` idea seems decent.

Reply:
What I mean is that before, we had the `shader.numStage` global accounting variable for the `stages` global array. Now we have the `numStage`global accounting variable  for the `stages` global array, I see no difference about risks of not being in sync (no less, no more).

The `numStage` variable was never part of the `stages` array.

And `numStage| was and is still global for the “currently processed shader” with the “currently processed stages list”, and there cannot be more than one global currently processed shader and more than one global current processed stage list, so that `numStage` global variable cant be storing data from something else than the `stages` global array.